### PR TITLE
Add scientific notation for BigDecimal formatting

### PR DESCRIPTION
**Pull request by fubhy**
Giulio is going to drive this over the finish line (property tests, fixes for `fromNumber`, etc.)

### CODE REVIEW

File: packages/effect/src/BigDecimal.ts [hunk 1]
```diff
@@ -811,20 +811,71 @@ export const fromBigInt = (n: bigint): BigDecimal => make(n, 0)
...
+ */
+export const fromNumber: (n: number) => BigDecimal = unsafeFromNumber
+
+/**
+ * Creates a `BigDecimal` from a `number` value.
+ *
+ * It is not recommended to convert a floating point number to a decimal directly,
+ * as the floating point representation may be unexpected.
+ *
+ * Returns `None` if the number is not finite.
```
Comment (gcanti) [inline packages/effect/src/BigDecimal.ts:852]:
```suggestion
 * Returns `None` if the number is not finite (`NaN`, `+Infinity` or `-Infinity`).
```

**Review by gcanti — CHANGES_REQUESTED** [review]

**Review by gcanti — APPROVED** [review]

**Comment by KhraksMamtsov** [comment]
sorry - but it seems we need to remember to process such a case - `fromString("0.0000e2e1")` now it returns `{ _id: 'BigDecimal', value: '0', scale: -16 }`

**Comment by fubhy** [comment]
> sorry - but it seems we need to remember to process such a case - `fromString("0.0000e2e1")` now it returns `{ _id: 'BigDecimal', value: '0', scale: -16 }`

That's an invalid input. When you say "we need to remember to process such a case" you mean "we need to throw in this case" right?

**Comment by KhraksMamtsov** [comment]
I think it is necessary to return `Option.none()` for any incorrect input (Correct me if I'm wrong)
In any case, it's probably worth reflecting this behavior in the tests.

**Comment by fubhy** [comment]
> I think it is necessary to return `Option.none()` for any incorrect input (Correct me if I'm wrong) In any case, it's probably worth reflecting this behavior in the tests.

Yes for sure. I was just trying to understand if you wanted a different kind of behavior (wasn't sure if you meant that this should be valid or not)

**Comment by fubhy** [comment]
@KhraksMamtsov Happy with this adjustment?

**Pull request merged**